README.md

# Flynn
`Flynn` is a vector graphics gaming framework for Javascript. It is the engine used on all the games
posted at www.vectoralchemy.com

[![Flynn](https://readthedocs.org/projects/flynn/badge/?version=latest)](http://flynn.readthedocs.io/en/latest/?badge=latest)  



## Overview

![logo][logo] 

The API is clean and the code is well organized, but there is (presently) not much end-user documentation.  You can find the current docs [here](http://flynn.readthedocs.io/en/latest).

I wrote Flynn to support my own game development and I've worked aggressively to push as much generalizable functionality into the engine as possible. Games implemented on top of Flynn end up having a light code base, and all browser issues (touch handling, window resizing, etc.) get handled by the engine. In practice, new game development stays very purely focused on game specifics & games run cross-platform "out of the box" with little effort. If you're mildly enterprising then the test app code and the code for the VA games should be sufficient for you to develop your own games on top of Flynn.  If there is developer interest then I'll start putting some more comprehensive documentation together.

![screen_flynn][screen_flynn] 
![screen_flynn_text][screen_flynn_text] 
![screen_flynn_box2d][screen_flynn_box2d] 
![screen_flynn_collision][screen_flynn_collision] 
![configuration][configuration] 
![code][code] 
![screen_polygondraw][screen_polygondraw] 

As of this writing (v2.5.1), Flynn supports the following features:

- Vector rendering
   - Draw Modes:
      - None (Lines only, no endpoint bloom)
      - Regular (Dimmer lines with endpoint bloom)
      - Thick (Regular, but thicker)
      - Flicker (Simulates vector display flicker)
   - World / Screen coordinate systems
   - Draw types:
      - Lines
      - Rects
      - Polygons
         - Rotation / Scaling
         - Multicolor
         - Collision detection
         - Pen-up (line discontinuities)
- Particle effects
- Timers
- Physics
   - Wrapper around Box2D, with vector rendering
- Font rendering
   - Justify Left/Center/Right
   - Arc text
   - Two fonts (Standard & Block)
- Controls
   - Laptop/Desktop
      - Keyboard binding
   - Virtual (touch) controls (on mobile devices)
      - Buttons
      - 8 position joystick
      - Analog joystick
      - D-Pad
- Configuration manager
   - Option selection
   - Save/Restore from cookies
   - Key configuration
- Score manager
   - Name entry
   - Score display
   - Save/Restore from cookies
- Projectiles
- Game State Manager
- Pacing management
   - Render FPS independent of execution rate
- Development tools
   - 20 FPS mode
   - Slow-motion mode
   - Frame rate monitor
   - Mousetouch (emulate touch on desktop machines)
- Canvas management
   - Browser resize
   - Screen rotation (mobile devices)


[logo]:  https://raw.githubusercontent.com/epmoyer/flynn/media/flynn_logo.png?raw=true "logo"
[screen_flynn]:  https://raw.githubusercontent.com/epmoyer/flynn/media/screen_flynn.png?raw=true&s=100 "screen_flynn"
[screen_flynn_text]: https://raw.githubusercontent.com/epmoyer/flynn/media/screen_flynn_text.png "screen_flynn_text"
[screen_flynn_collision]: https://raw.githubusercontent.com/epmoyer/flynn/media/screen_flynn_collision.png "screen_flynn_collision"
[screen_flynn_box2d]: https://raw.githubusercontent.com/epmoyer/flynn/media/screen_flynn_box2d.png "screen_flynn_box2d"
[configuration]:  https://raw.githubusercontent.com/epmoyer/flynn/media/flynn_configuration.png?raw=true "configuration"
[code]:  https://raw.githubusercontent.com/epmoyer/flynn/media/flynn_code-1024x674.png?raw=true "code"

[screen_polygondraw]: https://raw.githubusercontent.com/epmoyer/flynn/media/screen_polygondraw-1024x624.png "screen_polygondraw"

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres
to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## TODO
- TODO: 3.20.0 breaks invisible touch regions.  See TODO: in flynnInput.js.
- TODO: Generation of world_vertices (for 3D) does some of the same math as projection. Optimize performance.
- TODO: Update Flynn.Util.randomUnitV3() to have a normal distribution.
- TODO: Bring meshText back. Figure out how to reconcile with shatter.
- TODO: Require lifetimeTicks parameter in ._3DParticle init().

## Unreleased
**(None)**

## 3.22.0u 2021-06-06
### Added
- flynn3DParticles
    - Created Flynn._3DParticles

## 3.21.0u 2021-04-06
### Added
- flynn3D
    - Add backside culling.
    - Add opacity (hidden line removal).
    - Add `_3DMeshTetrahedron()` constructor.
- flynnText
    - Imported as-is; not reviewed yet.  Not added to demo/test app yet.

## 3.20.0 2021-03-02
### Fixed
- visible_states now work for touch regions and virtual (touch) joysticks (Flynn.mcp.input.addTouchRegion() and Flynn.mcp.input.addVirtualJoystick()).
### Changed
- **Breaking**: Touch regions are now disabled in game states not appearing in visible_states. This breaks legacy games that made use of invisible touch regions.  A future fix release will add an active_states argument which defaults to all states (thus restoring legacy functionality with no required code changes). 
### Added
- URL Flags
    - `quiethalt`: When set, halting (Accent/Tilde key in dev mode) will not display the "MCP HALTED. CLICK TO RESUME." banner.  This flag is provided to make it easier to get clean screenshots of a game for publishing.
- Optional dotSize argument to Flynn.Particles
- Flynn.InputHandler.flushButtons()
- Flynn.Util.randomShuffle()
- Ability to decompose meshes into individual line segments (and back).
    - Break meshes into segments with Flynn._3DMesh.to_segments()
    - Make meshes from segments with Flynn._3DMeshFromSegments()q
- screenOffsetV argument to Flynn._3DRenderer.

## 3.19.1 - 2020-12-12
### Added
- On-screen touch keyboard for high score name entry on touch devices (in flynnStateEnd).
- Add .getLastTouchLocation() to Flynn.InputHandler. Allows games to interrogate the position of touch events to implement their own full-screen touch handlers.
- Add "!" character to Flynn.Font.Block
- Remember high-score name (within the current session) and default to the previous name on subsequent high-score entry screens.
### Changed
- Flynn.Util.doBoundsBounce returns boolean (true if bounce occurred)
### Fixed
- Backspace key is now captured again (it used to work, but then broke with browser evolution).
- flynnCanvas was inspecting Game object for canvas size. Fixed to use Flynn.mcp object.
- flynnStateEnd will accept names for up to {leaderboard}.maxItems entries if Leaderboard not yet full.
- VA logo smoke restored.

## 3.18.0 - 2020-02-09
### Changed
- Engine
  - Adopt PixiJS v5.2.1

## 3.17.0 - 2019-12-08
### Added
- Engine
  - Add support for monochrome polygons to Flynn.Particles.shatter()
  - Flynn.Particles.shatter() **API changed**
    - Support min/max range for life, velocity, angular_velocity, and options for uniformly distributed exit angles.
- Test Application
  - Test monochrome polygon shatter.
  - Test new shatter options. Add "caret" ship.

### Fixed
- Engine
  - ctx.vectorText2 options opts.x, .y, and .angle now properly handle values of 0 (0 was getting overridden by null default).

## 3.19.0 - 2020-10-23
### Added
- Engine
  - Add "!" to Flynn.Font.Block.

## 3.16.0 - 2019-11-17
### Added
- Engine
  - transform_f argument to vectorText2 (callback function for vertex transformation).
- Test Application
  - Text transformation test/demo.

## 3.14.0 - 2019-08-10
### Added
- Engine
  - Lowercase characters now supported in Normal font.
  - Flynn.init() now takes forceUpperCase argument (defaults to true).
- Test Application
  - Show lowercase characters in ASCII table and sample text.

### Changed
- Engine
  - Improve some characters in Normal font: 'K', '(', ')'.

## 3.14.0 - 2019-07-13
### Fixed
- Engine
  - Flynn._3DRenderer.renderPoint() will now properly not render points which are behind the camera.
  - Buttons no longer get "stuck" DOWN if focus is switched away from the app while they are held. On restoration of focus, all pressed buttons are now forced into the UP state.
### Added
- Engine 
  - Flynn._3DMeshText() renders text to a 3D mesh object.
  - Flynn._3DMesh now has .pre_rotation
  - color alpha support
    - Flynn._3DMesh.init() now accepts alpha (defaults to 1.0)
    - ctx.vectoStart() now accepts alpha (defaults to 1.0)
    - Alpha is now supported in color values. Before alpha, vectors which used dark colors (grays, or dim colors) could be drawn atop brighter colors and obscure them, which does not match vector display behavior.
      - Colors can use either form:
        - "#RRGGBB"
        - "#RRGGBBAA"
      - Some standard Flynn colors (Flynn.Colors.[CYAN_DK, GRAY, GRAY_DK]) now use alpha.
  - Flynn.Util.heatmap()
  - Flynn.init() now accepts an optional disableUI parameter. 
    - If true the UI sounds will not be loaded.  Applications which do not use the Flynn UI can set disableUI, and omit the UI sounds from their site. 
    - NOTE: Without setting this option omitting the UI sounds will result in console errors being logged when Flynn attempts to load the UI sounds.
- Test Application
  - Add text meshes to 3D demo panel.
  - Add alpha color demos.

## 3.13.1 - 2019-01-11
### Fixed
- Engine
  - ctx.vectorText2 (and deprecated ctx.vectorText) now convert text argument to a string (so they can be used to display a number).

## 3.13.0 - 2018-10-09
### Changed
- Engine

flynnMain.js
// flynnMain
//------------
// This must be the first flynn script loaded

if (typeof Flynn == "undefined") {
  var Flynn = {};  // Create namespace
}

Flynn.DevPacingMode = {
	NORMAL:  0,
	SLOW_MO: 1,
	FPS_20:  2,
};

//------------
// Colors 
//------------
Flynn.Colors = {
	BLACK:        "#000000",
	BLUE:		  "#2020FF",
	WHITE:		  "#FFFFFF",
	GREEN:        "#00FF00",
	YELLOW:       "#FFFF00",
	RED:          "#FF0000",
	CYAN:         "#00FFFF",
	MAGENTA:	  "#FF00FF",
	CYAN_DK:      "#008080",
	
	ORANGE:       "#E55300",
	BROWN:        "#8B4513",
	YELLOW_DK:    "#808000",
	GRAY:         "#808080",
	GRAY_DK:      "#404040",

	LIGHTSKYBLUE: "#87CEFA",
	DODGERBLUE:   "#1E90FF",
	LIGHTBLUE:    "#ADD8E6",
};

//------------
// Keyboard input 
//------------
Flynn.KeyboardMap = {
	'0':        48,
	'1':        49,
	'2':        50,
	'3':        51,
	'4':        52,
	'5':        53,
	'6':        54,
	'7':        55,
	'8':        56,
	'9':        57,

	'a':        65,
	'b':        66,
	'c':        67,
	'd':        68,
	'e':        69,
	'f':        70,
	'g':        71,
	'h':        72,
	'i':        73,
	'j':        74,
	'k':        75,
	'l':        76,
	'm':        77,
	'n':        78,
	'o':        79,
	'p':        80,
	'q':        81,
	'r':        82,
	's':        83,
	't':        84,
	'u':        85,
	'v':        86,
	'w':        87,
	'x':        88,
	'y':        89,
	'z':        90,

	"'":        222,
	',':        188,
	'-':        189,
	'.':        190,
	'/':        191,
	';':        186,
	'=':        187,
	'[':        219,
	'\\':       220,
	']':        221,
	'`':        192,

	'tab':      9,
	'enter':    13,
	'shift':    16,
	'control':  17,
	'option':   18,
	'escape':   27,
	'spacebar': 32,
	'left':     37,
	'up':       38,
	'right':    39,
	'down':     40,
	'command':  91,

	'f1':       112,
	'f2':       113,
	'f3':       114,
	'f4':       115,
	'f5':       116,
	'f6':       117,
	'f7':       118,
	'f8':       119,
	'f9':       120,
	'f10':      121,
	'f11':      122,
	'f12':      123,

	'ICADE_up':    -1,
	'ICADE_down':  -2,
	'ICADE_left':  -3,
	'ICADE_right': -4,
	'ICADE_T1':    -5,
	'ICADE_T2':    -6,
	'ICADE_T3':    -7,
	'ICADE_T4':    -8,
	'ICADE_B1':    -9,
	'ICADE_B2':    -10,
	'ICADE_B3':    -11,
	'ICADE_B4':    -12,
};

//------------
// Vector Font 
//------------
Flynn.Font = {
	UnimplementedChar: [0,6,0,0,4,0,4,6,0,6,4,0,4,6,0,0,0,6],
	PenUp: 9999,
	CharacterHeight: 6,
	CharacterWidth: 4,
	CharacterGap: 2
};
Flynn.Font.CharacterSpacing = Flynn.Font.CharacterWidth + Flynn.Font.CharacterGap;
Flynn.Font.Points = {
	UNIMPLEMENTED_CHAR: Flynn.Font.UnimplementedChar,
	PEN_UP: Flynn.Font.PenUp,
	
	ASCII: [
		[1,0,2,5,1.5,5,1.5,6,2.5,6,2.5,5,2,5,3,0,1,0],               // !
		[1,0,1,1,Flynn.Font.PenUp,Flynn.Font.PenUp,3,0,3,1],         // "
		[1,5,1,1,1,2,0,2,4,2,3,2,3,1,3,5,3,                          // #
			4,4,4,0,4,1,4,1,5],
		[0,6,3,6,4,5,4,4,3,3,1,3,0,2,0,1,1,0,4,0,2,0,2,6],           // $        
		[0,6,4,0,2,3,2,1,0,1,0,3,4,3,4,5,2,5,2,3,0,6],               // %  
		[4,6,1,2,1,1,2,0,3,1,3,2,0,4,0,5,1,6,2,6,4,4],               // &  
		[2,0,2,1,2,0],                                               // '  
		[4,6,3,6,2,5,2,1,3,0,4,0],                                   // (  
		[1,0,2,0,3,1,3,5,2,6,1,6],                                   // )  
		[0,1,2,3,2,1,2,3,4,1,2,3,4,3,2,3,4,5,2,3,                    // *
			2,5,2,3,0,5,2,3,0,3],
		[2,5,2,1,2,3,0,3,4,3,2,3,2,5],                               // +  
		[2,5,1,6,2,5],                                               // ,  
		[1,3,3,3],                                                   // -  
		[1.5,6,1.5,5,2.5,5,2.5,6,1.5,6],                             // .  
		[1,6,3,0,1,6],                                               // /  
		[0,0,0,6,4,6,4,0,0,0],                                       // 0
		[2,0,2,6],                                                   // 1
		[0,0,4,0,4,3,0,3,0,6,4,6],                                   // 2
		[0,0,4,0,4,3,0,3,4,3,4,6,0,6],                               // 3
		[0,0,0,3,4,3,4,0,4,6],                                       // 4
		[4,0,0,0,0,3,4,3,4,6,0,6],                                   // 5
		[0,0,0,6,4,6,4,3,0,3],                                       // 6
		[0,0,4,0,4,6],                                               // 7
		[0,3,4,3,4,6,0,6,0,0,4,0,4,3],                               // 8
		[4,3,0,3,0,0,4,0,4,6],                                       // 9
		[2,1,1.5,1.5,2,2,2.5,1.5,2,1,                                // :
			Flynn.Font.PenUp, Flynn.Font.PenUp,
			2,4,1.5,4.5,2,5,2.5,4.5,2,4],
		[2,1,1.5,1.5,2,2,2.5,1.5,2,1,                                // ;
			Flynn.Font.PenUp, Flynn.Font.PenUp,
			1.5,4,2.5,4,1.5,6,1.5,4],
		[4,1,0,3,4,5,0,3,4,1],                                       // <  
		[0,1.5,4,1.5,Flynn.Font.PenUp,Flynn.Font.PenUp,0,4.5,4,4.5], // = 
		[0,1,4,3,0,5,4,3,0,1],                                       // >  
		[0,2,0,1,1,0,3,0,4,1,4,2,2,3,2,6,2,3,4,2,4,
			1,3,0,1,0,0,1,0,2],                                      // ?  
		[3,4,3,2,1,2,1,4,4,4,4,2,3,1,1,1,0,2,0,4,1,5,3,5],           // @
		[0,6,0,2,2,0,4,2,4,4,0,4,4,4,4,6],                           // A
		[0,3,0,6,2,6,3,5,3,4,2,3,0,3,0,0,2,0,3,1,3,2,2,3],           // B
		[4,0,0,0,0,6,4,6],                                           // C
		[0,0,0,6,2,6,4,4,4,2,2,0,0,0],                               // D
		[4,0,0,0,0,3,3,3,0,3,0,6,4,6],                               // E
		[4,0,0,0,0,3,3,3,0,3,0,6],                                   // F
		[4,2,4,0,0,0,0,6,4,6,4,4,2,4],                               // G
		[0,0,0,6,0,3,4,3,4,0,4,6],                                   // H
		[0,0,4,0,2,0,2,6,4,6,0,6],                                   // I
		[4,0,4,6,2,6,0,4],                                           // J
		[3,0,0,3,0,0,0,6,0,3,3,6],                                   // K
		[0,0,0,6,4,6],                                               // L
		[0,6,0,0,2,2,4,0,4,6],                                       // M
		[0,6,0,0,4,6,4,0],                                           // N
		[0,0,4,0,4,6,0,6,0,0],                                       // O
		[0,6,0,0,4,0,4,3,0,3],                                       // P
		[0,0,0,6,2,6,3,5,4,6,2,4,3,5,4,4,4,0,0,0],                   // Q
		[0,6,0,0,4,0,4,3,0,3,1,3,4,6],                               // R
		[4,0,0,0,0,3,4,3,4,6,0,6],                                   // S
		[0,0,4,0,2,0,2,6],                                           // T
		[0,0,0,6,4,6,4,0],                                           // U
		[0,0,2,6,4,0],                                               // V
		[0,0,0,6,2,4,4,6,4,0],                                       // W
		[0,0,4,6,2,3,4,0,0,6],                                       // X
		[0,0,2,2,4,0,2,2,2,6],                                       // Y
		[0,0,4,0,0,6,4,6],                                           // Z
		[3,0,1,0,1,6,3,6],                                           // [  
		[1,0,3,6],                                                   // /  
		[1,0,3,0,3,6,1,6],                                           // ]  
		[1,1,2,0,3,1],                                               // ^ 
		[0,6,4,6],                                                   // _
		[1.5,0,2.5,1],                                               // `
	],
};